swiftsimio/visualisation/projection.py

```diff
@@ -231,13 +232,34 @@ def project_pixel_grid(
     else:
         combined_mask = mask
 
-    common_arguments = dict(
-        x=(x[combined_mask] - x_min) / max_range,
-        y=(y[combined_mask] - y_min) / max_range,
-        m=m[combined_mask],
-        h=hsml[combined_mask] / max_range,
-        res=resolution,
-    )
+    xfinal = array((x[combined_mask] - x_min) / max_range)
+    yfinal = array((y[combined_mask] - y_min) / max_range)
+    mfinal = array(m[combined_mask])
+    hfinal = array(hsml[combined_mask] / max_range)
+    rescaled_box = array([box_x / max_range, box_y / max_range])
+
+    xall = xfinal.copy()
+    yall = yfinal.copy()
+    mall = mfinal.copy()
+    hall = hfinal.copy()
+    for xshift in [-1, 0, 1]:
+        for yshift in [-1, 0, 1]:
+            if xshift == 0 and yshift == 0:
+                continue
+            thisx = xfinal + xshift * rescaled_box[0]
+            thisy = yfinal + yshift * rescaled_box[1]
+            inside = (
+                (thisx - xshift * hfinal <= rescaled_box[0])
+                & (thisx - xshift * hfinal >= 0.0)
+                & (thisy - yshift * hfinal <= rescaled_box[1])
+                & (thisy - yshift * hfinal >= 0.0)
```

Comment:
dx = thisx - xshift * hfinal
dy = thisy - yshift * hfinal

Is np.logical_and.reduce() faster?

Reply:
Yes, I should precompute dx and dy.

According to the line profiler, `logical_and.reduce()` is actually slower.

Reply:
What about (x & x) & (y & y)? Or is this not faster... Maybe python is smart enough to reduce like that